lib/AST/Module.cpp

```diff
@@ -1811,36 +1809,18 @@ void SourceFile::createReferencedNameTracker() {
 }
 
 OpaqueTypeDecl *
-SourceFile::lookupOpaqueResultType(StringRef MangledName,
-                                   LazyResolver *resolver) {
-  // Check already-validated decls.
-  auto found = ValidatedOpaqueReturnTypes.find(MangledName);
-  if (found != ValidatedOpaqueReturnTypes.end())
+SourceFile::lookupOpaqueResultTypeByName(StringRef MangledName) {
+  auto found = OpaqueReturnTypesByName.find(MangledName);
+  if (found != OpaqueReturnTypesByName.end())
     return found->second;
-    
-  // If there are unvalidated decls with opaque types, go through and validate
-  // them now.
-  if (resolver && !UnvalidatedDeclsWithOpaqueReturnTypes.empty()) {
-    while (!UnvalidatedDeclsWithOpaqueReturnTypes.empty()) {
-      ValueDecl *decl = *UnvalidatedDeclsWithOpaqueReturnTypes.begin();
-      UnvalidatedDeclsWithOpaqueReturnTypes.erase(decl);
-      // FIXME(InterfaceTypeRequest): Remove this.
-      (void)decl->getInterfaceType();
-    }
-    
-    found = ValidatedOpaqueReturnTypes.find(MangledName);
-    if (found != ValidatedOpaqueReturnTypes.end())
-      return found->second;
-  }
-  
+
   // Otherwise, we don't have a matching opaque decl.
   return nullptr;
 }
 
-void SourceFile::markDeclWithOpaqueResultTypeAsValidated(ValueDecl *vd) {
-  UnvalidatedDeclsWithOpaqueReturnTypes.erase(vd);
+void SourceFile::registerDeclWithOpaqueResultType(ValueDecl *vd) {
   if (auto opaqueDecl = vd->getOpaqueResultTypeDecl()) {
```

Comment:
Has `getOpaqueResultTypeDecl` been request-ified? Will this work if `vd` hasn't been type checked yet, so its opaque return type decl doesn't yet exist?

Reply:
Hm, that does complicate this.  I think the right approach would be 

- Requestify `getOpaqueResultTypeDecl()`/`getOpaqueResultType()`
- Have the `OpaqueReturnTypesByName` table map to the parent declarations.
- Force `getOpaqueResultTypeDecl()` on lookup - this'll happen automatically if we compute the interface type

Reply:
That sounds good.